### PR TITLE
docs(versioning): changed 'extend' to 'widen' as a supported strategy

### PR DESCRIPTION
**Pull request by Gabriel-Ladzaretti**
## Changes
Rename 'extend' into 'widen' for all package managers.

## Context
related to issue #6241

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository
- [x] None

### CODE REVIEW

**Review by viceice — CHANGES_REQUESTED** [review]
This needs a migration to not break existing users. You also need to check code references.

**Comment by rarkins** [comment]
> This needs a migration to not break existing users. You also need to check code references.

This field was maybe just used for cosmetic/docs reasons

**Comment by viceice** [comment]
> > This needs a migration to not break existing users. You also need to check code references.
> 
> This field was maybe just used for cosmetic/docs reasons

So `widen` is used by code and we just had wrong docs?

**Comment by Gabriel-Ladzaretti** [comment]
few examples:
https://github.com/renovatebot/renovate/blob/564d0fec9eedd5faefae5158535d0167dab656aa/lib/versioning/ruby/index.ts#L121-L127
https://github.com/renovatebot/renovate/blob/564d0fec9eedd5faefae5158535d0167dab656aa/lib/versioning/npm/range.ts#L81-L83
https://github.com/renovatebot/renovate/blob/564d0fec9eedd5faefae5158535d0167dab656aa/lib/versioning/conan/range.ts#L253-L259

File: lib/versioning/cargo/index.ts [hunk 1]
```diff
@@ -9,7 +9,7 @@ export const urls = [
   'https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html',
 ];
 export const supportsRanges = true;
-export const supportedRangeStrategies = ['bump', 'extend', 'pin', 'replace'];
+export const supportedRangeStrategies = ['bump', 'widen', 'pin', 'replace'];
```
Comment (viceice) [inline lib/versioning/cargo/index.ts:12]:
```suggestion
export const supportedRangeStrategies: RangeStrategy = ['bump', 'widen', 'pin', 'replace'];
```
import from `lib/types`

https://github.com/renovatebot/renovate/blob/564d0fec9eedd5faefae5158535d0167dab656aa/lib/types/versioning.ts#L1

Please add that to all those `supportedRangeStrategies` definitions. That way we can enforce correct values.

Comment (Gabriel-Ladzaretti) [reply lib/versioning/cargo/index.ts:12]: TS2322: Type 'string[]' is not assignable to type 'RangeStrategy'.

Comment (viceice) [reply lib/versioning/cargo/index.ts:12]: Add a `as RangeStrategy`

File: lib/versioning/cargo/index.ts [hunk 2]
```diff
@@ -9,7 +10,12 @@ export const urls = [
   'https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html',
 ];
 export const supportsRanges = true;
-export const supportedRangeStrategies = ['bump', 'extend', 'pin', 'replace'];
+export const supportedRangeStrategies: RangeStrategy[] = [
+  'bump',
+  'widen',
+  'pin',
+  'replace',
+] as RangeStrategy[];
```
Comment (rarkins) [inline lib/versioning/cargo/index.ts:18]:
```suggestion
];
```

You need to revert all the other `as` statements too as they result in the arrays *not* being validated.

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/versioning/cargo/index.ts (hunk 2 above)
Comment (viceice) [reply lib/versioning/cargo/index.ts:18]: @rarkins Whitout this compilation fails 🤔

Comment (rarkins) [reply lib/versioning/cargo/index.ts]: Strange. But with it, there's no type validation?

Comment (viceice) [reply lib/versioning/cargo/index.ts]: Not sure. I've committed your suggestion to check ci.

File: lib/versioning/composer/index.ts [hunk 3]
```diff
@@ -14,13 +15,13 @@ export const urls = [
...
-export const supportedRangeStrategies = [
+export const supportedRangeStrategies: RangeStrategy[] = [
   'bump',
-  'extend',
+  'widen',
   'pin',
   'replace',
   'update-lockfile',
-];
+] as RangeStrategy[];
```
Comment (viceice) [inline lib/versioning/composer/index.ts:24]:
```suggestion
];
```

it seems to work without cast on ci

File: lib/versioning/hashicorp/index.ts [hunk 4]
```diff
@@ -8,7 +9,12 @@ export const urls = [
   'https://www.terraform.io/docs/configuration/terraform.html#specifying-a-required-terraform-version',
 ];
 export const supportsRanges = true;
-export const supportedRangeStrategies = ['bump', 'extend', 'pin', 'replace'];
+export const supportedRangeStrategies: RangeStrategy[] = [
+  'bump',
+  'widen',
+  'pin',
+  'replace',
+] as RangeStrategy[];
```
Comment (viceice) [inline lib/versioning/hashicorp/index.ts:17]:
```suggestion
];
```

File: lib/versioning/hex/index.ts [hunk 5]
```diff
@@ -6,7 +7,12 @@ export const id = 'hex';
 export const displayName = 'Hex';
 export const urls = ['https://hexdocs.pm/elixir/Version.html'];
 export const supportsRanges = true;
-export const supportedRangeStrategies = ['bump', 'extend', 'pin', 'replace'];
+export const supportedRangeStrategies: RangeStrategy[] = [
+  'bump',
+  'widen',
+  'pin',
+  'replace',
+] as RangeStrategy[];
```
Comment (viceice) [inline lib/versioning/hex/index.ts:15]:
```suggestion
];
```

File: lib/versioning/ivy/index.ts [hunk 6]
```diff
@@ -17,7 +18,12 @@ export const id = 'ivy';
 export const displayName = 'Ivy';
 export const urls = ['https://ant.apache.org/ivy/'];
 export const supportsRanges = true;
-export const supportedRangeStrategies = ['bump', 'extend', 'pin', 'replace'];
+export const supportedRangeStrategies: RangeStrategy[] = [
+  'bump',
+  'widen',
+  'pin',
+  'replace',
+] as RangeStrategy[];
```
Comment (viceice) [inline lib/versioning/ivy/index.ts:26]:
```suggestion
];
```

File: lib/versioning/maven/index.ts [hunk 7]
```diff
@@ -22,7 +23,12 @@ export const urls = [
   'https://maven.apache.org/enforcer/enforcer-rules/versionRanges.html',
 ];
 export const supportsRanges = true;
-export const supportedRangeStrategies = ['bump', 'extend', 'pin', 'replace'];
+export const supportedRangeStrategies: RangeStrategy[] = [
+  'bump',
+  'widen',
+  'pin',
+  'replace',
+] as RangeStrategy[];
```
Comment (viceice) [inline lib/versioning/maven/index.ts:31]:
```suggestion
];
```

File: lib/versioning/npm/index.ts [hunk 8]
```diff
@@ -12,7 +13,12 @@ export const urls = [
   'https://semver.npmjs.com/',
 ];
 export const supportsRanges = true;
-export const supportedRangeStrategies = ['bump', 'extend', 'pin', 'replace'];
+export const supportedRangeStrategies: RangeStrategy[] = [
+  'bump',
+  'widen',
+  'pin',
+  'replace',
+] as RangeStrategy[];
```
Comment (viceice) [inline lib/versioning/npm/index.ts:21]:
```suggestion
];
```

File: lib/versioning/pep440/index.ts [hunk 9]
```diff
@@ -1,12 +1,18 @@
...
 export const displayName = 'PEP440';
 export const urls = ['https://www.python.org/dev/peps/pep-0440/'];
 export const supportsRanges = true;
-export const supportedRangeStrategies = ['bump', 'extend', 'pin', 'replace'];
+export const supportedRangeStrategies: RangeStrategy[] = [
+  'bump',
+  'widen',
+  'pin',
+  'replace',
+] as RangeStrategy[];
```
Comment (viceice) [inline lib/versioning/pep440/index.ts:15]:
```suggestion
];
```

File: lib/versioning/poetry/index.ts [hunk 10]
```diff
@@ -14,7 +16,12 @@ export const id = 'poetry';
 export const displayName = 'Poetry';
 export const urls = ['https://python-poetry.org/docs/versions/'];
 export const supportsRanges = true;
-export const supportedRangeStrategies = ['bump', 'extend', 'pin', 'replace'];
+export const supportedRangeStrategies: RangeStrategy[] = [
+  'bump',
+  'widen',
+  'pin',
+  'replace',
+] as RangeStrategy[];
```
Comment (viceice) [inline lib/versioning/poetry/index.ts:24]:
```suggestion
];
```

File: lib/versioning/rez/index.ts [hunk 11]
```diff
@@ -23,7 +25,12 @@ export const id = 'rez';
 export const displayName = 'rez';
 export const urls = ['https://github.com/nerdvegas/rez'];
 export const supportsRanges = true;
-export const supportedRangeStrategies = ['bump', 'extend', 'pin', 'replace'];
+export const supportedRangeStrategies: RangeStrategy[] = [
+  'bump',
+  'widen',
+  'pin',
+  'replace',
+] as RangeStrategy[];
```
Comment (viceice) [inline lib/versioning/rez/index.ts:33]:
```suggestion
];
```

File: lib/versioning/ruby/index.ts [hunk 12]
```diff
@@ -22,7 +23,12 @@ export const urls = [
   'https://www.devalot.com/articles/2012/04/gem-versions.html',
 ];
 export const supportsRanges = true;
-export const supportedRangeStrategies = ['bump', 'extend', 'pin', 'replace'];
+export const supportedRangeStrategies: RangeStrategy[] = [
+  'bump',
+  'widen',
+  'pin',
+  'replace',
+] as RangeStrategy[];
```
Comment (viceice) [inline lib/versioning/ruby/index.ts:31]:
```suggestion
];
```

File: lib/versioning/swift/index.ts [hunk 13]
```diff
@@ -8,7 +9,12 @@ export const id = 'swift';
 export const displayName = 'Swift';
 export const urls = ['https://swift.org/package-manager/'];
 export const supportsRanges = true;
-export const supportedRangeStrategies = ['bump', 'extend', 'pin', 'replace'];
+export const supportedRangeStrategies: RangeStrategy[] = [
+  'bump',
+  'widen',
+  'pin',
+  'replace',
+] as RangeStrategy[];
```
Comment (viceice) [inline lib/versioning/swift/index.ts:17]:
```suggestion
];
```

**Review by viceice — DISMISSED** [review]

**Review by rarkins — CHANGES_REQUESTED** [review]
Can you sure that every `versioning` we have has a typed `supportedRangeStrategies: RangeStrategy[]` export? Some are missing in this PR.

**Comment by Gabriel-Ladzaretti** [comment]
> Can you sure that every versioning we have has a typed supportedRangeStrategies: RangeStrategy[] export? Some are missing in this PR.

added three more, pretty sure this is it.

**Review by viceice — APPROVED** [review]

**Comment by viceice** [comment]
> > Can you sure that every versioning we have has a typed supportedRangeStrategies: RangeStrategy[] export? Some are missing in this PR.
> 
> added three more, pretty sure this is it.

https://github.com/renovatebot/renovate/blob/main/.github/contributing.md#re-requesting-a-review

File: lib/versioning/gradle/index.ts [hunk 14]
```diff
@@ -17,7 +18,7 @@ export const urls = [
   'https://docs.gradle.org/current/userguide/single_versions.html#version_ordering',
 ];
 export const supportsRanges = true;
-export const supportedRangeStrategies = ['pin'];
+export const supportedRangeStrategies: RangeStrategy[] = ['pin'];
```
Comment (rarkins) [inline lib/versioning/gradle/index.ts:21]: This seems strange, but I suppose at worst it's a docs error

Comment (viceice) [reply lib/versioning/gradle/index.ts:21]: https://github.com/renovatebot/renovate/blob/main/lib/versioning/gradle/index.ts#L194-L197

I think that is right. 🤷‍♂️

**Review by rarkins — APPROVED** [review]

**Pull request merged**